packages/compass-import-export/src/csv/csv-utils.ts

```diff
@@ -11,116 +11,212 @@ import {
+    // ObjectId and UUID both have toHexString() which does exactly what we want
+    return value.toHexString();
+  }
+
+  if (bsonType === 'Binary') {
+    return EJSON.stringify(value.value());
+  }
+
+  if (bsonType === 'BSONRegExp') {
+    const bsonregex = value as BSONRegExp;
+    return formatCSVValue(`/${bsonregex.pattern}/${bsonregex.options}`, {
+      delimiter,
+    });
+  }
+
+  if (bsonType === 'Decimal128') {
+    // This should turn it into a number string with exponent
+    return value.toString();
+  }
+
+  if (bsonType === 'Timestamp') {
+    // This should turn it into a number string
+    return value.toString();
+  }
+
+  if (bsonType === 'MinKey') {
+    // Same as mongoexport
+    return '$MinKey';
+  }
+
+  if (bsonType === 'MaxKey') {
+    // Same as mongoexport
+    return '$MaxKey';
+  }
+
+  // Symbol, Code, DBRef and whatever new types get added
+  return formatCSVValue(EJSON.stringify(value), { delimiter });
+}
+
+export function csvHeaderNameToFieldName(name: string) {
+  return name.replace(/\[\d+\]/g, '');
+}
 
 const MIN_INT = -2147483648;
 const MAX_INT = 2147483647;
 const MIN_LONG = BigInt('-9223372036854775808');
 const MAX_LONG = BigInt('9223372036854775807');
 // from papaparse: https://github.com/mholt/PapaParse/blob/aa0046865f1b4e817ebba6966d6baf483e0652d7/papaparse.js#L1024
-const FLOAT = /^\s*-?(\d+\.?|\.\d+|\d+\.\d+)([eE][-+]?\d+)?\s*$/;
+const FLOAT_REGEX = /^\s*-?(\d+\.?|\.\d+|\d+\.\d+)([eE][-+]?\d+)?\s*$/;
 // from papaparse: https://github.com/mholt/PapaParse/blob/aa0046865f1b4e817ebba6966d6baf483e0652d7/papaparse.js#L1025
-const ISO_DATE =
-  /^((\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z)))$/;
+const ISO_DATE_REGEX =
+  /^((\d{4}-[01]\d-[0-3]\d[T ][0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z)))$/;
+const DATEONLY_REGEX = /^\d{4}-[01]\d-[0-3]\d$/;
+// a regular expression for detecting regular expressions
+const REGEX_REGEX = /^\/.*\/\w*$/;
+// from js-bson: https://github.com/mongodb/js-bson/blob/5b837a9e5019016529a83700f3ba3065d5e53e80/src/objectid.ts#L6
+// this also supports mongoexport's format
+const OBJECTID_REGEX = /^(ObjectId\()?([0-9a-fA-F]{24})\)?$/;
+// from js-bson: https://github.com/mongodb/js-bson/blob/5b837a9e5019016529a83700f3ba3065d5e53e80/src/uuid_utils.ts#L5
+const UUID_REGEX =
+  /^(?:[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}|[0-9a-f]{12}4[0-9a-f]{3}[89ab][0-9a-f]{15})$/i;
 const TRUTHY_STRINGS = ['t', 'true', 'TRUE', 'True'];
 const FALSY_STRINGS = ['f', 'false', 'FALSE', 'False'];
 const NULL_STRINGS = ['Null', 'NULL', 'null'];
 
-export function detectFieldType(
+const parenthesis = {
+  '{': '}',
+  '[': ']',
+};
+
+function isEJSON(value: string) {
```

Comment:
Should we check if the value can be parsed as valid json? (`JSON.parse(value)` doesn't throw).  Or is that too much compute?

Reply:
Actually if there's a lot of stuff like that in the file it will still be really slow because each field would it make past those checks. And this is just meant to be heuristics - if it fails during import the parse error will be caught and logged and the user can still select something else manually.

Plus the column would have to look like EJSON for each row encountered for it to detect that which would also help weed out false positives.

So I think this is probably the right compromise for now. We can always make it stricter in a follow-up if it is ever a problem.

Reply:
Another compromise might be to add an option strict: true/false. Then when analyzing we can set it to false and the second detection that gets done at import time if the column is specified as Mixed can pass strict because then the stakes are a little bit higher if it gets it wrong.